### Example:

**Dataset**:

I'm using a sample dataset with information about different regions, products, and the corresponding sales.

In [1]:
import pandas as pd

# Sample dataset
data = {'Region': ['East', 'West', 'East', 'West', 'East'],
        'Product': ['A', 'B', 'A', 'B', 'C'],
        'Sales': [100, 200, 150, 300, 120]}

df = pd.DataFrame(data)
df


Region Product  Sales
0   East       A    100
1   West       B    200
2   East       A    150
3   West       B    300
4   East       C    120

***Grouping by a Single Column***

Task: Group by Region and calculate total sales.

In [2]:
# Grouping by Region and summing Sales
grouped_data = df.groupby('Region')['Sales'].sum()
grouped_data


Region
East    370
West    500
Name: Sales, dtype: int64

Analysis:

East Region: Total sales are 370.

West Region: Total sales are 500.


***Grouping by a Single Column with Multiple Aggregations***

Task: Apply both sum() and mean() to Sales for each region.

In [3]:
# Applying sum and mean together
agg_data = df.groupby('Region')['Sales'].agg(['sum', 'mean'])
agg_data


sum        mean
Region                 
East    370  123.333333
West    500  250.000000

Analysis:

East Region: Total sales are 370 with an average of 123.33.

West Region: Total sales are 500 with an average of 250.00.


***Grouping by Multiple Columns***

Task: Group by Region and Product, then calculate both sum() and mean() for Sales.

In [4]:
# Grouping by Region and Product, and applying multiple aggregations
grouped_by_multiple = df.groupby(['Region', 'Product'])['Sales'].agg(['sum', 'mean'])
grouped_by_multiple


sum   mean
Region Product            
East   A        250  125.0
       C        120  120.0
West   B        500  250.0

Analysis:

East Region, Product A: Total sales are 250 with an average of 125.

East Region, Product C: Total sales are 120 with an average of 120.

West Region, Product B: Total sales are 500 with an average of 250.


# Exercise: Sales Analysis for RetailCo

You’ve been hired as a data analyst for RetailCo, a company selling products across different regions. The company wants to understand how sales are distributed by region, product category, and customer segment. They also want insights into which regions and product categories are generating the highest sales.

Tasks:
1. Analyze Sales by Region:

 * Group sales data by regions and calculate total and average sales.
 * Identify the region with the highest and lowest total sales.
2. Analyze Sales by Product Category:

 *  Group by product categories and calculate total and average sales.
 * Determine which product category is performing best.

3. Analyze Sales by Customer Segment:

 * Group sales data by customer segments (e.g., Regular vs. Premium customers).
 * Calculate the total number of sales for each customer segment.

4. Combined Analysis (Multiple Columns):

 * Group by Region and Product Category to see which combinations have the highest sales.
 * Apply multiple aggregations (e.g., total and average sales) in the same grouping.



Bonus Challenge:

Create a pivot table that shows total sales by Region and Customer Segment.
Identify any trends based on customer segments in different regions.


Imaginary Dataset:
Here's a small fictional dataset for this exercise.

In [2]:
import pandas as pd

data = {
    'Region': ['East', 'West', 'East', 'South', 'West', 'East', 'South', 'North', 'North', 'West'],
    'Product_Category': ['Electronics', 'Clothing', 'Clothing', 'Furniture', 'Electronics', 'Furniture', 'Clothing', 'Electronics', 'Furniture', 'Clothing'],
    'Customer_Segment': ['Regular', 'Premium', 'Regular', 'Regular', 'Premium', 'Premium', 'Regular', 'Premium', 'Regular', 'Regular'],
    'Sales': [200, 450, 120, 300, 500, 400, 150, 600, 350, 250]
}

df = pd.DataFrame(data)
df


Region Product_Category Customer_Segment  Sales
0   East      Electronics          Regular    200
1   West         Clothing          Premium    450
2   East         Clothing          Regular    120
3  South        Furniture          Regular    300
4   West      Electronics          Premium    500
5   East        Furniture          Premium    400
6  South         Clothing          Regular    150
7  North      Electronics          Premium    600
8  North        Furniture          Regular    350
9   West         Clothing          Regular    250

In [6]:
#1. Analyze Sales by Region:

 #* Group sales data by regions and calculate total and average sales.
 #* Identify the region with the highest and lowest total sales.
agg_data =df.groupby('Region')['Sales'].agg(['sum','mean'])
agg_data

sum   mean
Region             
East     720  240.0
North    950  475.0
South    450  225.0
West    1200  400.0

In [9]:
#2. Analyze Sales by Product Category:

 #*  Group by product categories and calculate total and average sales.
 #* Determine which product category is performing best.
analise2=df.groupby('Product_Category')['Sales'].agg(['sum','mean'])
analise2

sum        mean
Product_Category                  
Clothing           970  242.500000
Electronics       1300  433.333333
Furniture         1050  350.000000

In [10]:
#Analyze Sales by Customer Segment:

# * Group sales data by customer segments (e.g., Regular vs. Premium customers).
# * Calculate the total number of sales for each customer segment.
analise3=df.groupby('Customer_Segment')['Sales'].sum()
analise3

Customer_Segment
Premium    1950
Regular    1370
Name: Sales, dtype: int64

In [14]:
#4. Combined Analysis (Multiple Columns):

# * Group by Region and Product Category to see which combinations have the highest sales.
 #* Apply multiple aggregations (e.g., total and average sales) in the same grouping.
analise4=df.groupby(['Region', 'Product_Category'])['Sales'].agg(['sum','mean'])
analise4



sum   mean
Region Product_Category            
East   Clothing          120  120.0
       Electronics       200  200.0
       Furniture         400  400.0
North  Electronics       600  600.0
       Furniture         350  350.0
South  Clothing          150  150.0
       Furniture         300  300.0
West   Clothing          700  350.0
       Electronics       500  500.0